autonat/autonat-v2.md

```diff
@@ -0,0 +1,193 @@
+# AutonatV2: spec
+
+
+| Lifecycle Stage | Maturity                 | Status | Latest Revision |
+|-----------------|--------------------------|--------|-----------------|
+| 1A              | Working Draft            | Active | r2, 2023-04-15  |
+
+Authors: [@sukunrt]
+
+Interest Group: [@marten-seemann], [@marcopolo], [@mxinden]
+
+[@sukunrt]: https://github.com/sukunrt
+[@marten-seemann]: https://github.com/marten-seemann
+[@mxinden]: https://github.com/mxinden
+[@marcopolo]: https://github.com/marcopolo
+
+
+## Overview
+
+A priori, a node cannot know if it is behind a NAT / firewall or if it is
+publicly reachable. Knowing its NAT status is essential for the node to be
+well-behaved in the network: A node that's behind a NAT / firewall doesn't need
+to advertise its (undialable) addresses to the rest of the network, preventing
+superfluous dials from other peers. Furthermore, it might actively seek to
+improve its connectivity by finding a relay server, which would allow other
+peers to establish a relayed connection.
+
+In `autonat v2` client sends a priority ordered list of addresses. On receiving
+this list the server dials the first address on the list that it is capable of
+dialing. `autonat v2` allows nodes to determine reachability for individual
+addresses. Using `autonat v2` nodes can build an address pipeline where they can
+test individual addresses discovered by different sources like identify, upnp
+mappings, circuit addresses etc for reachability. Having a priority ordered list
+of addresses provides the ability to verify low priority addresses.
+Implementations can generate low priority address guesses and add them to
+requests for high priority addresses as a nice to have. This is especially
+helpful when introducing a new transport. Initially, such a transport will not
+be widely supported in the network. Requests for verifying such addresses can be
+reused to get information about other addresses
+
+Compared to `autonat v1` there are two major differences
+1. `autonat v1` allowed testing reachability for the node. `autonat v2` allows
+testing reachability for an individual address
+2. `autonat v2` provides a mechanism for nodes to verify whether the peer
+actually successfully dialled an address.
+
+
+## AutoNAT V2 Protocol
```

Comment:
This section would benefit from a sequence diagram that shows what is exchanged on which stream at what point in time.

I've also made a naming suggestion further down.